### PR TITLE
iox-#1144 set condition variable parameter and prepare configuration via CMake

### LINKED ISSUE #1144: Adjust `MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE` to 256
## Brief feature description

When a user would like to attach a large number of subscriber/clients/servers to a listener or waitset the number of `MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE` has to be increased. This should be configurable via a cmake switch like `IOX_MAX_PUBLISHERS`.

Furthermore the variables:
* `MAX_NUMBER_OF_ATTACHMENTS_PER_WAITSET`
* `MAX_NUMBER_OF_EVENTS_PER_LISTENER`

are tightly coupled (identical) to the notifiers per condition variable. It would make sense to remove those two and use in the listener and waitset only `MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE`.

### PR DESCRIPTION
**Pull request by dkroenke**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

- extend deployment types with parameter for number of condition variables
- fix clang-tidy warnings in posh types

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [ ] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1144

### CODE REVIEW

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 1]
```diff
@@ -100,14 +100,13 @@ namespace popo
...
-constexpr uint32_t MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE = 128U;
-constexpr uint32_t MAX_NUMBER_OF_ATTACHMENTS_PER_WAITSET = 128U;
-static_assert(MAX_NUMBER_OF_ATTACHMENTS_PER_WAITSET <= MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE,
-              "The WaitSet capacity is restricted by the maximum amount of notifiers per condition variable.");
-// Listener
-constexpr uint8_t MAX_NUMBER_OF_EVENTS_PER_LISTENER = 128U;
-static_assert(MAX_NUMBER_OF_EVENTS_PER_LISTENER <= MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE,
-              "The Listener capacity is restricted by the maximum amount of notifiers per condition variable.");
+
+constexpr uint32_t MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE =
```
Comment (dkroenke) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:104]: I tend to re-use the name because it is quite generic, a comment below should clarify that this is used by the Listener and Waitset API.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 2]
```diff
@@ -35,7 +35,7 @@ struct iox_ws_storage_t_
 {
     // the value of the array size is the result of the following formula:
     // sizeof(WaitSet) / 8
-    uint64_t do_not_touch_me[2965];
+    uint64_t do_not_touch_me[5958];
```
Comment (dkroenke) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:38]: @elfenpiff Here the new values for the types in C-Binding

Comment (elfenpiff) [reply iceoryx_binding_c/include/iceoryx_binding_c/types.h:38]: Ouh no, this is evil ... I think we may need some macro magic.

Maybe we need a different approach here. One can also define via cmake macros which are then known in all c/cpp files and I think we require this approach here. In `iceoryx_posh_types.hpp` the variable is then set to this macro value and the size of the waitset/listener are then calculated based on that macro value.

If we do not do this the c binding may crash when the user increases the value. If we introduce a max number we gain nothing in the c binding since here we will use then always the worst case number. I think the macro value approach is the only approach which solves those problems - what do you think?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:104]: Maybe we can shorten it a bit to `MAX_NUMBER_OF_NOTIFIERS`? The info that it belongs to a condition variable is not really important for the user.

File: iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_BINDING_C_DEPLOYMENT_H
+#define IOX_BINDING_C_DEPLOYMENT_H
+
+// clang-format off
+
+#define WAITSET_STORAGE_SIZE (@IOX_MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE@ * 24)
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in]: Nope, you forgot to take the size of the index type into account, which must be adjusted for alignment reasons ;)

File: iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl [hunk 4]
```diff
@@ -61,31 +67,157 @@ inline cxx::expected<ListenerError> Listener::attachEvent(
...
+template <uint64_t Capacity>
 template <typename T>
-inline void Listener::detachEvent(T& eventOrigin) noexcept
+inline void ListenerImpl<Capacity>::detachEvent(T& eventOrigin) noexcept
 {
     NotificationAttorney::disableEvent(eventOrigin);
 }
 
-inline constexpr uint64_t Listener::capacity() noexcept
+template <uint64_t Capacity>
```
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl:91]: The code in listener.inl was only moved without change from listener.cpp. This was required since the listener is now a template class.

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 5]
```diff
@@ -100,14 +100,12 @@ namespace popo
 using WaitSetIsConditionSatisfiedCallback = cxx::ConstMethodCallback<bool>;
 }
 constexpr uint32_t MAX_NUMBER_OF_CONDITION_VARIABLES = 1024U;
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:102]: I'm wondering if this should also be configurable?

File: iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl (hunk 4 above)
Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl:91]: The code in listener.inl was only moved without change from listener.cpp. This was required since the listener is now a template class.

Comment (elfenpiff) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/listener.inl:91]: The code in listener.inl was only moved without change from listener.cpp. This was required since the listener is now a template class.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 6]
```diff
@@ -18,8 +18,21 @@
...
 
+#include "iceoryx_binding_c/iceoryx_binding_c_deployment.h"
 #include "internal/c2cpp_binding.h"
 
+#if defined(__APPLE__)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    144 + numberOfAttachments * 168 - 8 * (((numberOfAttachments + 1) / 2) - 1)
+#elif defined(_WIN32)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    168 + numberOfAttachments * 176 - 8 * (((numberOfAttachments + 1) / 2) - 1)
```
Comment (elBoberido) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:29]: I wonder why this is more complicated for Windows and mac

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 5 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:102]: Maybe not a bad idea. This restricts the user on how many waitset/listeners can be created in an iceoryx environment

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp [hunk 7]
```diff
@@ -34,6 +34,60 @@ extern "C" {
...
+    EXPECT_THAT(sizeof(ListenerImpl<16>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(16)));
+    EXPECT_THAT(sizeof(ListenerImpl<17>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(17)));
+
+    EXPECT_THAT(sizeof(ListenerImpl<99>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(99)));
+    EXPECT_THAT(sizeof(ListenerImpl<100>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(100)));
+    EXPECT_THAT(sizeof(ListenerImpl<101>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(101)));
+
+    EXPECT_THAT(sizeof(ListenerImpl<999>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(999)));
+    EXPECT_THAT(sizeof(ListenerImpl<1000>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(1000)));
+    EXPECT_THAT(sizeof(ListenerImpl<1001>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(1001)));
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:61]: Thats why I included 8 and 16 but alright ...

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 5 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:102]: But I would keep this for iceoryx 3.0 and for the iceoryx posh types refactoring.

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:65]: But 8 and 16 fit into the same fixed width type. 255 and 256 not.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 8]
```diff
@@ -18,8 +18,21 @@
...
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    (144 + numberOfAttachments * 168 - 8 * (((numberOfAttachments + 1) / 2) - 1))
+#elif defined(_WIN32)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    (168 + numberOfAttachments * 176 - 8 * (((numberOfAttachments + 1) / 2) - 1))
+#else
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments) (((128 + numberOfAttachments * 140) / 8) * 8)
+#endif
+
+#define CALCULATE_STORAGE_SIZE_FOR_WAITSET(numberOfAttachments) (552 + numberOfAttachments * 184)
```
Comment (elBoberido) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:34]: Counter proposal to this mess #1169

Comment (elfenpiff) [reply iceoryx_binding_c/include/iceoryx_binding_c/types.h]: I find the proposal is the right way to go, but not for 2.0 - not enough time and it is uncertain that we encounter some weird surprises here.

We have tests in place to handle it cleanly and provide a working solution for the user so I would pursue this approach for now.

**Comment by elfenpiff** [comment]
Since I worked on this PR as well I would exclude myself from approving this.

File: iceoryx_posh/include/iceoryx_posh/popo/listener.hpp [hunk 9]
```diff
@@ -27,13 +27,50 @@
 #include "iceoryx_posh/popo/notification_attorney.hpp"
 #include "iceoryx_posh/popo/notification_callback.hpp"
 #include "iceoryx_posh/popo/trigger_handle.hpp"
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/popo/listener.hpp:30]: ~~Why is this include needed?~~ Should be moved to `.inl`

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp [hunk 10]
```diff
@@ -100,14 +100,12 @@ namespace popo
...
-constexpr uint32_t MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE = 128U;
-constexpr uint32_t MAX_NUMBER_OF_ATTACHMENTS_PER_WAITSET = 128U;
-static_assert(MAX_NUMBER_OF_ATTACHMENTS_PER_WAITSET <= MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE,
-              "The WaitSet capacity is restricted by the maximum amount of notifiers per condition variable.");
-// Listener
-constexpr uint8_t MAX_NUMBER_OF_EVENTS_PER_LISTENER = 128U;
-static_assert(MAX_NUMBER_OF_EVENTS_PER_LISTENER <= MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE,
-              "The Listener capacity is restricted by the maximum amount of notifiers per condition variable.");
+
+constexpr uint32_t MAX_NUMBER_OF_NOTIFIERS = build::IOX_MAX_NUMBER_OF_NOTIFIERS;
```
Comment (mossmaurice) [inline iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:104]: Why was the change only done for `popo::Listener` and not for `popo::WaitSet`?

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp [hunk 11]
```diff
@@ -34,6 +34,76 @@ extern "C" {
 
 #include <iostream>
 
+inline void verifyStorageSizeCalculationForListener()
```
Comment (mossmaurice) [inline iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:37]: Those tests are a bit odd. Could we not split them up for the different integer sizes? E.g. `verifyStorageSizeCalculationForListenerUInt{8,16, ..}`

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 12]
```diff
@@ -18,8 +18,32 @@
...
+///            or SanityCheck.VerifyStorageSizeCalculationForWaitSet
+///        2. Take a look at the expected numbers of size 1 (A1) and 2 (A2).
+///        3. Find the parameters m, n for the function StorageSize(x) = m + n * x
+///        4. Rerun the the tests to verify if the parameters are correct.
+#if defined(__APPLE__)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    (144 + numberOfAttachments * 168 - 8 * (((numberOfAttachments + 1) / 2) - 1))
+#elif defined(_WIN32)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    (168 + numberOfAttachments * 176 - 8 * (((numberOfAttachments + 1) / 2) - 1))
```
Comment (mossmaurice) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:36]: What are those magic numbers for apple and windows? Please document them!

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 13]
```diff
@@ -18,8 +18,32 @@
...
+#include "iceoryx_binding_c/iceoryx_binding_c_deployment.h"
 #include "internal/c2cpp_binding.h"
 
+/// @brief if the function parameters change due to an update of the listener or
+///        waitset recalculate them with the following approach.
+///        1. Run SanityCheck.SanityCheck.VerifyStorageSizeCalculationForListener
+///            or SanityCheck.VerifyStorageSizeCalculationForWaitSet
+///        2. Take a look at the expected numbers of size 1 (A1) and 2 (A2).
+///        3. Find the parameters m, n for the function StorageSize(x) = m + n * x
+///        4. Rerun the the tests to verify if the parameters are correct.
```
Comment (mossmaurice) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:30]:
```suggestion
///        4. Re-run the the tests to verify if the parameters are correct.
```

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 14]
```diff
@@ -18,8 +18,32 @@
...
+    (144 + numberOfAttachments * 168 - 8 * (((numberOfAttachments + 1) / 2) - 1))
+#elif defined(_WIN32)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
+    (168 + numberOfAttachments * 176 - 8 * (((numberOfAttachments + 1) / 2) - 1))
+#else
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments) (((128 + numberOfAttachments * 140) / 8) * 8)
+#endif
+
+#if defined(_WIN32)
+#define CALCULATE_STORAGE_SIZE_FOR_WAITSET(numberOfAttachments) (552 + numberOfAttachments * 168)
```
Comment (mossmaurice) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:42]: Again, I see lots of magic number. Please describe them with a few words.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h (hunk 12 above)
Comment (elfenpiff) [reply iceoryx_binding_c/include/iceoryx_binding_c/types.h:36]: @mossmaurice See the comment above the macros which explains how the magic numbers were extracted.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h (hunk 14 above)
Comment (elfenpiff) [reply iceoryx_binding_c/include/iceoryx_binding_c/types.h:42]: @mossmaurice See the comment above the macros.

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp (hunk 11 above)
Comment (elfenpiff) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:37]: Here I am testing if the size calculation function is correct by picking multiple values which have to map to the underlying linear function.

What would be the benefit of using different integer sizes? `sizeof` always returns an `uint64_t` and the `ListenerImpl` capacity is also always the same integer size?

File: iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp (hunk 10 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp:104]: I do not understand? This number is applied to waitset and listener attachments two lines below?

The goal is to make this configurable via cmake switch.

File: iceoryx_posh/include/iceoryx_posh/popo/listener.hpp (hunk 9 above)
Comment (elfenpiff) [reply iceoryx_posh/include/iceoryx_posh/popo/listener.hpp:30]: I disagree. The `inl` file is the header implementation and adding there headers hides this from the user which is including this header file.
Every header which is required to use this should be listed here and not secretly hidden in the `inl` file.

Now we need a third one to decide the tie.

Mysterious stranger please add a thumbs up to one of the remarks to break the tie.

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 15]
```diff
@@ -18,8 +18,32 @@
...
 
+/// @brief if the function parameters change due to an update of the listener or
+///        waitset recalculate them with the following approach.
+///        1. Run SanityCheck.SanityCheck.VerifyStorageSizeCalculationForListener
+///            or SanityCheck.VerifyStorageSizeCalculationForWaitSet
+///        2. Take a look at the expected numbers of size 1 (A1) and 2 (A2).
+///        3. Find the parameters m, n for the function StorageSize(x) = m + n * x
+///        4. Re-run the the tests to verify if the parameters are correct.
+#if defined(__APPLE__)
+#define CALCULATE_STORAGE_SIZE_FOR_LISTENER(numberOfAttachments)                                                       \
```
Comment (MatthiasKillat) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:32]: I do not believe any of these formulas will do us any good since they assume too much about how the compiler will structure the data. It is true that there must be a lower bound, but theoretically there does not have to be an upper bound as the compiler is allowed to leave empty bytes (alignment) etc. for whatever reason. It will of course try to conserve space and minimize these empty bytes, but we just cannot know the exact number.

It is hard to get a (close) upper bound, and in general not possible without compiler specific knowledge.

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp [hunk 16]
```diff
@@ -34,6 +34,76 @@ extern "C" {
...
+
+    EXPECT_THAT(sizeof(ListenerImpl<255>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(255)));
+    EXPECT_THAT(sizeof(ListenerImpl<256>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(256)));
+    EXPECT_THAT(sizeof(ListenerImpl<257>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(257)));
+
+    EXPECT_THAT(sizeof(ListenerImpl<999>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(999)));
+    EXPECT_THAT(sizeof(ListenerImpl<1000>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(1000)));
+    EXPECT_THAT(sizeof(ListenerImpl<1001>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(1001)));
+
+    EXPECT_THAT(sizeof(ListenerImpl<2047>), Eq(CALCULATE_STORAGE_SIZE_FOR_LISTENER(2047)));
```
Comment (MatthiasKillat) [inline iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:67]: Is the user able to set these template argument for the ListenerImpl? If not, it is mainly important that it works for the configured value and no number of finite checks will be able to tell us whether it will work for all possible values. You assume there is a linear function (and it likely is), but it does not have to be. The compiler may do something special if it gets larger than page size etc.

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:65]: I am not sure that using powers of two (+1/-1) is really solving the fundamental problem of the assumptions of the formula.

File: iceoryx_posh/include/iceoryx_posh/popo/listener.hpp (hunk 9 above)
Comment (MatthiasKillat) [reply iceoryx_posh/include/iceoryx_posh/popo/listener.hpp:30]: .hpp and .inl work differently from .hpp and .cpp insofar that the user is aware of all dependencies (as the compiler has to be). We should probably not make it harder for the user to find them, but on the other hand consider.

If we think of them as all pasted in one file (which is what the preprocessor does), where would we include the headers? While one can  include the headers directly before use (at the latest moment), it is highly unusual. I am not aware that it will cause any errors, for the compiler it should not make any difference. But it makes it harder for the user to understand the dependencies.

I am fine with either, but would like to avoid implementation specific headers in the file the user includes. This does not really lead to nice structure (IMO) with .inl files.

In the end I dislike the .inl file approach and prefer the detail/internal namespace approach as it also reduces namespace pollution with implementation details. Then we can have an implementation header using the detail sub namespace. This header would now include the implementation specific headers. If one is to imitate this with .inl I think we would include the implementation specific header in the .inl file.

TLDR: if we want to mimic what is usually done with detail namespaces we would include the header in the .inl file. However, it is not as clean as having the detail header and detail sub namespace.

File: iceoryx_posh/include/iceoryx_posh/popo/listener.hpp [hunk 17]
```diff
@@ -197,21 +203,31 @@ class Listener
...
-            m_loffliStorage[LoFFLi::requiredIndexMemorySize(MAX_NUMBER_OF_EVENTS_PER_LISTENER) / sizeof(uint32_t)];
+        LoFFLi::Index_t m_loffliStorage[LoFFLi::requiredIndexMemorySize(Capacity) / sizeof(uint32_t)];
         LoFFLi m_loffli;
         std::atomic<uint64_t> m_indicesInUse{0U};
     } m_indexManager;
 
 
     std::thread m_thread;
-    concurrent::smart_lock<Event_t, std::recursive_mutex> m_events[MAX_NUMBER_OF_EVENTS_PER_LISTENER];
+    concurrent::smart_lock<internal::Event_t, std::recursive_mutex> m_events[Capacity];
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/popo/listener.hpp:213]: Regarding my comment with the detail namespace: it appears we already have something like a detail namespace here (internal) and also some kind of implementation class `ListenerImpl`.

File: iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in [hunk 18]
```diff
@@ -16,6 +16,7 @@
 #ifndef IOX_BINDING_C_DEPLOYMENT_H
 #define IOX_BINDING_C_DEPLOYMENT_H
 
-#define IOX_BUILD_GENERATED_MAX_NUMBER_OF_NOTIFIERS @IOX_MAX_NUMBER_OF_NOTIFIERS@
+/// @note must be equal to MAX_NUMBER_OF_NOTIFIERS in iceoryx_posh_types.hpp
```
Comment (elfenpiff) [inline iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in:19]: Could you please also add a hint here which files one has to touch when this is going to be activated.
* `iceoryx_posh/cmake/IceoryxPoshDeployment.cmake`
* `iceoryx_posh/cmake/iceoryx_posh_deployment.hpp.in `
* `iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp`

Comment (dkroenke) [reply iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in:19]: Comment added

File: iceoryx_binding_c/include/iceoryx_binding_c/types.h [hunk 19]
```diff
@@ -18,8 +18,32 @@
 #ifndef IOX_BINDING_C_TYPES_H
 #define IOX_BINDING_C_TYPES_H
 
+#include "iceoryx_binding_c/iceoryx_binding_c_deployment.h"
 #include "internal/c2cpp_binding.h"
 
+/// @brief if the function parameters change due to an update of the listener or
+///        waitset recalculate them with the following approach.
+///        1. Run SanityCheck.SanityCheck.VerifyStorageSizeCalculationForListener
```
Comment (elBoberido) [inline iceoryx_binding_c/include/iceoryx_binding_c/types.h:26]: Too much `SanityCheck` ;)

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:65]: No, but it is one of the things which might change the size of the structures

File: iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp (hunk 16 above)
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp:67]: We have checks for the compiled version. It is in `iceoryx_binding_c/test/moduletests/test_types_storage_size.hpp`. This checks are just there to check the formula.

File: iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_BINDING_C_DEPLOYMENT_H
+#define IOX_BINDING_C_DEPLOYMENT_H
+
+/// @note must be equal to MAX_NUMBER_OF_NOTIFIERS in iceoryx_posh_types.hpp
+/// @todo The following files need to be touched when activating the CMake configuration of IOX_MAX_NUMBER_OF_NOTIFIERS
+/// enable constexpr IOX_MAX_NUMBER_OF_NOTIFIERS in iceoryx_posh/cmake/iceoryx_posh_deployment.hpp.in
+/// enable IOX_MAX_NUMBER_OF_NOTIFIERS parameter in iceoryx_posh/cmake/IceoryxPoshDeployment.cmake
+/// use build::IOX_MAX_NUMBER_OF_NOTIFIERS parameter in iceoryx_posh/include/iceoryx_posh/iceoryx_posh_types.hpp
+#define IOX_BUILD_GENERATED_MAX_NUMBER_OF_NOTIFIERS 256 /// @IOX_MAX_NUMBER_OF_NOTIFIERS@
```
Comment (elBoberido) [inline iceoryx_binding_c/cmake/iceoryx_binding_c_deployment.h.in:24]: This can be forgotten when the number is changed in iceoryx_posh. Could you maybe create a `IOX_INTERNAL_MAX_NUMBER_OF_NOTIFIERS` cmake variable which will not be set/overridden with `-DIOX_...` and use that one?

File: doc/website/release-notes/iceoryx-v2-0-0.md [hunk 21]
```diff
@@ -44,6 +44,9 @@ servers (`iox::popo::MessagingPattern::REQ_RES`) [\#27](https://github.com/eclip
     - Limitations
         - The port introspection is not aware of the new `Client` and `Server` [\#1128](https://github.com/eclipse-iceoryx/iceoryx/issues/1128)
         - The DDS gateway is not aware of the new `Server` [\#1145](https://github.com/eclipse-iceoryx/iceoryx/issues/1145)
+- Let `MAX_NUMBER_OF_NOTIFIERS_PER_CONDITION_VARIABLE` be configurable via CMake switch [\#1144](https://github.com/eclipse-iceoryx/iceoryx/issues/1144)
+- Let `MAX_NUMBER_OF_NOTIFIERS` be configurable via CMake switch [\#1144](https://github.com/eclipse-iceoryx/iceoryx/issues/1144)
```
Comment (elfenpiff) [inline doc/website/release-notes/iceoryx-v2-0-0.md]: I think this is no longer valid since you commented this out?

Comment (dkroenke) [reply doc/website/release-notes/iceoryx-v2-0-0.md]: Yes, that is fixed, you will see it when you refresh the page

**Review by elBoberido — DISMISSED** [review]
Not ideal, but I guess it is the best we can do without delaying v2.0

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_binding_c/test/moduletests/main.cpp [hunk 22]
```diff
@@ -32,6 +32,16 @@ TEST(SanityCheck, CheckStorageSizeAndAlingment)
...
 
+TEST(SanityCheck, VerifyStorageSizeCalculationForListener)
+{
+    verifyStorageSizeCalculationForListener();
+}
+
+TEST(SanityCheck, VerifyStorageSizeCalculationForWaitSet)
+{
+    verifyStorageSizeCalculationForWaitSet();
+}
```
Comment (elBoberido) [inline iceoryx_binding_c/test/moduletests/main.cpp:43]: please add test ids

**Review by elBoberido — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
